source/secure_tunneling.c

```diff
@@ -128,6 +130,96 @@ static int s_secure_tunneling_close(struct aws_secure_tunnel *secure_tunnel) {
     return AWS_OP_SUCCESS;
 }
 
+static void s_secure_tunneling_on_send_data_complete_callback(
+    struct aws_websocket *websocket,
+    int error_code,
+    void *user_data) {
+    UNUSED(websocket);
+    struct data_tunnel_pair *pair = user_data;
+    pair->secure_tunnel->config.on_send_data_complete(error_code, user_data);
+}
+
+static bool s_secure_tunneling_send_data_call(
+    struct aws_websocket *websocket,
+    struct aws_byte_buf *out_buf,
+    void *user_data) {
+    UNUSED(websocket);
+    struct data_tunnel_pair *pair = user_data;
+    struct aws_byte_buf *buffer = &pair->buf;
+    AWS_RETURN_ERROR_IF2(aws_byte_buf_write(out_buf, buffer->buffer, buffer->len - 1) == AWS_OP_SUCCESS, false);
+    aws_byte_buf_clean_up(buffer);
+    aws_mem_release(pair->secure_tunnel->config.allocator, (void *)pair);
+    return true;
+}
+
+static void s_init_websocket_send_frame_options(
+    struct aws_websocket_send_frame_options *frame_options,
+    struct data_tunnel_pair *pair) {
+    frame_options->payload_length = pair->buf.len - 1;
+    frame_options->user_data = pair;
+    frame_options->stream_outgoing_payload = s_secure_tunneling_send_data_call;
+    frame_options->on_complete = s_secure_tunneling_on_send_data_complete_callback;
+    frame_options->opcode = AWS_WEBSOCKET_OPCODE_BINARY;
+    frame_options->fin = true;
+    frame_options->high_priority = false;
+    AWS_ZERO_STRUCT(frame_options->rsv);
+}
+
+static int s_secure_tunneling_send(
+    struct aws_secure_tunnel *secure_tunnel,
+    const struct aws_byte_buf *data,
+    enum aws_iot_st_message_type type) {
+    struct aws_iot_st_msg message;
+    message.streamId = secure_tunnel->stream_id;
+    message.ignorable = 0;
+    message.type = type;
+    if (data != NULL) {
+        message.payload.allocator = data->allocator;
+        message.payload.buffer = data->buffer;
+        message.payload.capacity = data->capacity;
+        message.payload.len = data->len;
+    } else {
+        message.payload = aws_byte_buf_from_c_str("");
+    }
+    struct data_tunnel_pair *pair =
+        (struct data_tunnel_pair *)aws_mem_acquire(secure_tunnel->config.allocator, sizeof(struct data_tunnel_pair));
+    pair->secure_tunnel = secure_tunnel;
+    AWS_RETURN_ERROR_IF2(
```

Comment:
I'm inferring that this macro handles returning error if the enclosed call is an error. On error you will leak *pair

Reply:
+1

Reply:
I removed them and am doing the same goto cleanup method that i used in the serializer